src/metabase/feature_extraction/comparison.clj

```diff
@@ -1,48 +1,16 @@
 (ns metabase.feature-extraction.comparison
   "Feature vector similarity comparison."
   (:require [bigml.histogram.core :as h.impl]
+            [clojure.math.numeric-tower :as num]
             [clojure.set :as set]
-            [kixi.stats
-             [core :as stats]
-             [math :as math]]
+            [kixi.stats.core :as stats]
             [metabase.feature-extraction
              [feature-extractors :as fe]
-             [histogram :as h]]
+             [histogram :as h]
+             [math :as math]]
             [redux.core :as redux])
   (:import com.bigml.histogram.Histogram))
 
-(def magnitude
-  "Transducer that claclulates magnitude (Euclidean norm) of given vector.
-   https://en.wikipedia.org/wiki/Euclidean_distance"
-  (redux/post-complete (redux/pre-step + math/sq) math/sqrt))
-
-(defn cosine-distance
-  "Cosine distance between vectors `a` and `b`.
-   https://en.wikipedia.org/wiki/Cosine_similarity"
-  [a b]
-  (transduce identity
-             (redux/post-complete
-              (redux/fuse {:magnitude-a (redux/pre-step magnitude first)
-                           :magnitude-b (redux/pre-step magnitude second)
-                           :product     (redux/pre-step + (partial apply *))})
-              (fn [{:keys [magnitude-a magnitude-b product]}]
-                (some->> (fe/safe-divide product magnitude-a magnitude-b)
-                         (- 1))))
-             (map (comp (partial map double) vector) a b)))
-
-(defn head-tails-breaks
-  "Pick out the cluster of N largest elements.
-   https://en.wikipedia.org/wiki/Head/tail_Breaks"
```

Comment:
These functions for example are 100x easier to understand because of the docstrings. 👍